### DISCUSSION TITLE
How to refresh table every 5 seconds

### DISCUSSION THREAD

**Discussion by Shivar90** [zauberzeug/nicegui#2558] [Q&A]
### Question

I want to refresh the only table every 5 seconds. I read about nicegui documentation about "https://nicegui.io/documentation/refreshable#refreshable_ui_with_parameters" and "https://nicegui.io/documentation/timer#call_a_function_after_a_delay". Which is better approach to refresh the table. 
I shared my MRE code .Please nicegui community give any input and comment on this.
```Python
from nicegui import ui, events
import json
import requests
import pandas as pd


def my_table():
    # Define columns
    columns = [
        {'name': 'action', 'label': 'Action', 'field': 'action', 'required': True, 'align': 'left'},
        {'name': 'Group', 'label': 'Group', 'field': 'Group', 'required': True, 'align': 'left'},
    ]

    rows = [
        {'id': None, 'action': False, 'Group': 'A'},
        {'id': None, 'action': True, 'Group': 'B'},
        {'id': None, 'action': True, 'Group': 'C'},
        {'id': None, 'action': True, 'Group': 'D'},
    ]
    # Create the table with defined columns, add df data to table
    table = ui.table(columns=columns, rows=rows, selection=None, pagination=5).classes('w-full')
    Columns_Name = table.columns
    print("Columns Name:", Columns_Name)
    if any(column['label'] == 'Action' for column in table.columns):
        print(f"The table contains the 'Action' column.")
        table.add_slot('body-cell-action', '''
            <q-td key="action" :props="props">
                <q-toggle 
                    :color="props.value ? 'red' : 'green'" 
                    keep-color
                    :label="props.value ? 'Get' : 'Leave'" 
                    checked-icon="clear" 
                    v-model="props.row.action" 
                    v-model="check_value" 
                    @update:model-value="() => $parent.$emit(
                        <!-- callback function -->
                        'action_event', <!-- should match string passed to `table.on` below -->
                        
                        <!-- arguments, can be whatever you want really -->
                        props.row,   <!-- this would pass the whole row of data -->
                        props.row.Group,   <!-- this would pass data of the Group column for this row -->
                        props.value ? 'Get' : 'Leave'  <!-- this will pass a string indicating the action based on the toggle switch state -->
                    )"
                />           
            </q-td>
        ''')
    else:
        print(f"The table does not contain the 'Action' column.")
    with table.add_slot('top-right'):
        with ui.input(placeholder='Search').props('type=search').bind_value(table, 'filter').add_slot('append'):
            ui.icon('search')
    ui.timer(5.0, lambda: table, active=True)


def handle_action_event(e: events.GenericEventArguments) -> None:
    print(f"{e.sender}")  # this will be the table object
    print(f"{e.args}")  # this will be the args defined in the emit call
    ui.notify(f'{e.args[1]} wants action {e.args[2]}')
    for row in rows:
        if row['id'] == e.args['id']:
            row['name'] = e.args['name']
    ui.notify(f'Table.rows is now: {table.rows}')


ui.timer(1.0, lambda: my_table(), once=True)
ui.run(port=8080)
```

**Comment by rodja** [comment]
You can use a `ui.timer` to call `.refresh()` on a refreshable every 5 sec.

**Reply by Shivar90** [reply]
please give any minimum code how to implement that you mean use refresh inside ui.timer as callback.

**Comment by falkoschindler** [comment]
Using `ui.timer` and `.refresh()`:

```py
@ui.refreshable
def my_table():
    columns = [{'name': 'time', 'label': 'Time', 'field': 'time'}]
    rows = [{'time': time.time()}]
    ui.table(columns=columns, rows=rows)

my_table()

ui.timer(5.0, my_table.refresh)
```

Or you simply update the `rows` without re-creating the whole table:

```py
columns = [{'name': 'time', 'label': 'Time', 'field': 'time'}]
rows = [{'time': time.time()}]
table = ui.table(columns=columns, rows=rows)

def update_rows():
    table.rows = [{'time': time.time()}]

ui.timer(5.0, update_rows)
```

**Reply by Shivar90** [reply]
Thanks @falkoschindler and @rodja able to update table using @ui.refreshable ,even using button created in qusar "q-toggle button" framework.

**Reply by Shivar90** [reply]
@falkoschindler and @rodja I have a user authentication implemented in backend server where restrictive Api's require user authentication token.
The table on my screen has a data coming from the rest Api which requires this user authentication token.
 
When I test my application with multiple user login from different systems, data is coming correctly in the table and rest api calls are using different token for each Api call.
 
But when I add above mentioned refresh table mechanism in my code, the authentication token is getting mismatched.
For example: If two users are doing operations on the table toggle button on different rows, still authentication token is getting mismatch during the request.
Even though 2nd user is making a toggle action, but server shows it is getting updated from user 1.
 
I have implemented user authentication in the backend server, where restricted APIs require a user authentication token. 
The data displayed in the table on my screen is retrieved from REST APIs, each of which requires this user authentication token.
 
During testing, I observed that when I log in with multiple users from different systems, the data is correctly displayed in the table, and REST API calls are utilizing different tokens for each call.
 
However, an issue arises when I introduce above mentioned mechanism to refresh the table in my code. 
In such cases, authentication tokens are getting mismatched.
 
For instance, if two users perform operations on the table, such as toggling buttons on different rows, the authentication tokens end up being mismatched.  Refer attached table image. 
Even though the second user is triggering a toggle action, the server incorrectly identifies it as an update from the first user.

For toggle action in table, please refer above MRE code function "def handle_action_event(e: events.GenericEventArguments)"

<img width="914" alt="Table_ToogleButton" src="https://github.com/zauberzeug/nicegui/assets/36196330/533934d7-a57e-4c1e-9166-fa589b7ab271">
I have used "app.storage.user" for updating and getting the token. Referred below example:
 https://github.com/zauberzeug/nicegui/blob/main/examples/authentication/main.py" 
 
I have used page builder function "@ui.page('/')" as mentioned in documentation "https://nicegui.io/documentation/storage#storage" 

 I would appreciate any insights or suggestions on resolving this synchronization issue.

**Reply by falkoschindler** [reply]
@Shivar90 This problem seems mostly unrelated to your original question. Please open a new issue or Q&A and - if possible - try to reduce it to a minimal example. I guess the table with all its complexity like slots isn't needed. Shaving off unnecessary complexity is a crucial first step for analyzing the problem and allows us to help much more efficiently. Thanks!
